opm/simulators/linalg/cuistl/CuSparseMatrix.hpp

```diff
@@ -0,0 +1,302 @@
+#include <opm/simulators/linalg/cuistl/detail/CuSparseHandle.hpp>
+#include <opm/simulators/linalg/cuistl/detail/safe_conversion.hpp>
+#include <vector>
+
+namespace Opm::cuistl
+{
+
+/**
+ * @brief The CuSparseMatrix class simple wrapper class for a CuSparse matrix.
+ *
+ * @note we currently only support simple raw primitives for T (double and float). Block size is handled through the
+ * block size parameter
+ *
+ * @tparam T the type to store. Can be either float, double or int.
+ *
+ * @note we only support square matrices.
+ *
+ * @note We only support Block Compressed Sparse Row Format (BSR) for now.
+ */
+template <typename T>
+class CuSparseMatrix
+{
+public:
+    //! Create the sparse matrix specified by the raw data.
+    //!
+    //! \note Prefer to use the constructor taking a const reference to a matrix instead.
+    //!
+    //! \param[in] nonZeroElements the non-zero values of the matrix
+    //! \param[in] rowIndices      the row indices of the non-zero elements
+    //! \param[in] columnIndices   the column indices of the non-zero elements
+    //! \param[in] numberOfNonzeroElements number of nonzero elements
+    //! \param[in] blockSize size of each block matrix (typically 3)
+    //! \param[in] numberOfRows the number of rows
+    //!
+    //! \note We assume numberOfNonzeroBlocks, blockSize and numberOfRows all are representable as int due to
+    //!       restrictions in the current version of cusparse. This might change in future versions.
+    CuSparseMatrix(const T* nonZeroElements,
+                   const int* rowIndices,
+                   const int* columnIndices,
+                   size_t numberOfNonzeroBlocks,
+                   size_t blockSize,
+                   size_t numberOfRows);
+
+    /**
+     * We don't want to be able to copy this for now (too much hassle in copying the cusparse resources)
+     */
+    CuSparseMatrix(const CuSparseMatrix&) = delete;
+
+    /**
+     * We don't want to be able to copy this for now (too much hassle in copying the cusparse resources)
+     */
+    CuSparseMatrix& operator=(const CuSparseMatrix&) = delete;
+
+    virtual ~CuSparseMatrix();
+
+    /**
+     * @brief fromMatrix creates a new matrix with the same block size and values as the given matrix
+     * @param matrix the matrix to copy from
+     * @param copyNonZeroElementsDirectly if true will do a memcpy from matrix[0][0][0][0], otherwise will build up the
+     * non-zero elements by looping over the matrix. Note that setting this to true will yield a performance increase,
+     * but might not always yield correct results depending on how the matrix matrix has been initialized. If unsure,
+     * leave it as false.
+     * @tparam MatrixType is assumed to be a Dune::BCRSMatrix compatible matrix.
+     */
+    template <class MatrixType>
+    static CuSparseMatrix<T> fromMatrix(const MatrixType& matrix, bool copyNonZeroElementsDirectly = false);
```

Comment:
Why not make this a normal constructor instead of a static function?

Reply:
I would postpone refactoring when we get the cubuffer.